Snowflake.Data/Core/SFChunkDownloader.cs

```diff
@@ -60,38 +57,63 @@ class SFChunkDownloader
             FillDownloads();
         }
 
-        private BlockingCollection<Task<SFResultChunk>> _downloadTasks;
-        
-        private void FillDownloads()
-        {
-            _downloadTasks = new BlockingCollection<Task<SFResultChunk>>(prefetchSlot);
+        private BlockingCollection<Lazy<Task<SFResultChunk>>> _downloadTasks;
+        private ConcurrentQueue<Lazy<Task<SFResultChunk>>> _downloadQueue;
 
-            Task.Run(() =>
+        private void RunDownloads()
+        {
+            try
             {
-                foreach (var c in chunks)
+                while (_downloadQueue.TryDequeue(out var task) && !externalCancellationToken.IsCancellationRequested)
                 {
-                    _downloadTasks.Add(DownloadChunkAsync(new DownloadContext()
+                    if (!task.IsValueCreated)
                     {
-                        chunk = c,
-                        chunkIndex = nextChunkToDownloadIndex,
-                        qrmk = this.qrmk,
-                        chunkHeaders = this.chunkHeaders,
-                        cancellationToken = this.externalCancellationToken
-                    }));
+                        task.Value.Wait(externalCancellationToken);
+                    }
                 }
+            }
+            catch (Exception)
+            {
+                //Don't blow from background threads.
+            }
+        }
+
+
+        private void FillDownloads()
+        {
+            _downloadTasks = new BlockingCollection<Lazy<Task<SFResultChunk>>>();
+
+            foreach (var c in chunks)
+            {
+                var t = new Lazy<Task<SFResultChunk>>(() => DownloadChunkAsync(new DownloadContext()
+                {
+                    chunk = c,
+                    chunkIndex = c.ChunkIndex,
+                    qrmk = this.qrmk,
+                    chunkHeaders = this.chunkHeaders,
+                    cancellationToken = this.externalCancellationToken,
+                }));
+
+                _downloadTasks.Add(t);
+            }
+
+            _downloadTasks.CompleteAdding();
+
+            _downloadQueue = new ConcurrentQueue<Lazy<Task<SFResultChunk>>>(_downloadTasks);
+
+            for (var i = 0; i < prefetchSlot && i < chunks.Count; i++)
+                Task.Run(new Action(RunDownloads));

```

Comment:
It seems that prefetchSlot is used only when chunk downloader starts. However, I still think it make sense to prefetch when a chunk is taken from the BlockingCollections. That's why we use prefetchSlot for the upper-bound of blocking collections. i.e. always keeps 2 slots for prefetching.

Reply:
The prefetch slot is used to spawn tasks that will persist until all of the data is downloaded.

Reply:
So what's the point in your code using BlockingCollection? It's not blocking anything. All `Lazy<Task<SFResultChunk>>` element will be added into BlockingCollection at the beginning of the chunk downloader right? Then data will only be downloaded when main thread has finished consuming previous chunk and call `GetNextChunkAsync`

Reply:
I kept the BlockingCollection for the `IsCompleted` behavior, we could use a simple Queue, but we would then need to implement our own checks to prevent a race condition where the consumer hits an empty Queue before it has been populated and assume that it has completed.